src/lib.rs

```diff
@@ -572,67 +477,198 @@ impl KafkaJsonToDelta {
+    fn are_partition_offsets_match(&self, state: &mut ProcessingState) -> bool {
+        let mut result = true;
+        for (partition, offset) in &state.delta_partition_offsets {
+            let version = state
+                .delta_writer
+                .last_transaction_version(&self.app_id_for_partition(*partition));
+
+            if let Some(version) = version {
+                // if messages in kafka are consecutive then offset should always be `version+1` for
+                // safe commit, but since it's no guaranteed by kafka, we only check for `less than`.
+                if *offset != version {
                     info!(
-                        "Partition {} has no recorded offset. Not seeking consumer. Default starting offsets will be applied.",
-                        k
+                        "Conflict offset for partition {}: state={:?}, delta={:?}",
+                        partition, offset, version
                     );
+                    result = false;
                 }
             }
         }
+        result
+    }
 
+    async fn reset_state_guarded(
+        &self,
+        state: &mut ProcessingState,
+    ) -> Result<(), KafkaJsonToDeltaError> {
+        let mut partition_assignment = self.partition_assignment.lock().await;
+        self.reset_state(state, &mut partition_assignment).await?;
         Ok(())
     }
 
-    async fn consume_value_buffers(
+    async fn reset_state(
         &self,
         state: &mut ProcessingState,
-    ) -> Result<(Vec<Value>, HashMap<DataTypePartition, DataTypeOffset>), KafkaJsonToDeltaError>
-    {
-        let mut partition_assignment = self.partition_assignment.lock().await;
+        partition_assignment: &mut PartitionAssignment,
+    ) -> Result<(), KafkaJsonToDeltaError> {
+        state.delta_writer.reset()?;
+        state.value_buffers.reset();
+        state.delta_partition_offsets.clear();
 
-        let ConsumedBuffers {
-            values,
-            partition_offsets,
-        } = state.value_buffers.consume();
+        // assuming that partition_assignment has correct partitions as keys
+        let partitions: Vec<DataTypePartition> =
+            partition_assignment.assignment.keys().map(|p| *p).collect();
 
-        partition_assignment.update_offsets(&partition_offsets);
+        info!("Resetting state with partitions: {:?}", &partitions);
 
-        let partition_offsets = partition_assignment.partition_offsets();
+        // update offsets to the latest from the delta log
+        for partition in partitions.iter() {
+            let version = state
+                .delta_writer
+                .last_transaction_version(&self.app_id_for_partition(*partition));
+
+            if let Some(version) = version {
+                // transaction version in delta log is the latest offset stored for partition,
+                // so we need to seek to the incremented to get the next messages.
+                let offset = version + 1;
```

Comment:
I'm a bit concerned about this. Given that Kafka offsets are not guaranteed to be contiguous, I don't know what the behavior would be if we seek to e.g. `42`, but there is no `42` in Kafka (i.e. Kafka offsets jump from `40` to `43`).

Reply:
I'll do some follow up research, but one possible workaround might be to seek to the known offset and skip messages in the run loop until `>`

Reply:
either way, i'd like to hold the fix if we need one for a separate PR.

Reply:
curious when we get a rebalance event, what offsets do we get? do we get the offset for the last consumed message or the offset for the next message to consume?

Reply:
> if we seek to e.g. 42, but there is no 42 in Kafka 
Well, there's only way to find out for sure, gonna check it

Reply:
Okay so we need to fix that, opened and issue https://github.com/delta-io/kafka-delta-ingest/issues/29